crates/ide-diagnostics/src/handlers/remove_unnecessary_else.rs

```diff
@@ -387,4 +465,39 @@ fn test() {
 "#,
         );
     }
+
+    #[test]
+    fn no_diagnostic_if_not_expr_stmt() {
+        check_diagnostics_with_needless_return_disabled(
+            r#"
+fn test1() {
+    let _x = if a {
+        return;
+    } else {
+        1
+    };
+}
+
+fn test2() {
+    let _x = if a {
+        return;
+    } else if b {
+        return;
+    } else if c {
+        1
+    } else {
+        return;
+    };
+}
+"#,
+        );
+        check_diagnostics_with_disabled(
+            r#"
+fn test3() {
+    foo(if a { return 1 } else { 0 })
+}
+"#,
+            std::iter::once("E0308".to_owned()),
+        );
```

Comment:
@Veykril is there a better test for this case?

Reply:
That should be fine, but let's add a return type to `test3` so we don't have to supress the type mismatch diagnostic